src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java

```diff
@@ -0,0 +1,208 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceSplitUtil.class);
+
+  /**
+   * Generates and returns Salesforce splits for a query
+   * @param query the query for the sobject
+   * @param bulkConnection used to create salesforce jobs
+   * @param enablePKChunk indicates if pk chunking is enabled
+   * @return list of salesforce splits
+   */
+  public static List<SalesforceSplit> getQuerySplits(String query, BulkConnection bulkConnection,
+                                                     boolean enablePKChunk) {
+    return Stream.of(getBatches(query, bulkConnection, enablePKChunk))
+      .map(batch -> new SalesforceSplit(batch.getJobId(), batch.getId(), query))
+      .collect(Collectors.toList());
+  }
+
+  /**
+   * Based on query length sends query to Salesforce to receive array of batch info. If query is within limit, executes
+   * original query. If not, switches to wide object logic, i.e. generates Id query to retrieve batch info for Ids only
+   * that will be used later to retrieve data using SOAP API.
+   *
+   * @param query SOQL query
+   * @param bulkConnection bulk connection
+   * @param enablePKChunk enable PK Chunking
+   * @return array of batch info
+   */
+  private static BatchInfo[] getBatches(String query, BulkConnection bulkConnection, boolean enablePKChunk) {
+    try {
+      if (!SalesforceQueryUtil.isQueryUnderLengthLimit(query)) {
+        LOG.debug("Wide object query detected. Query length '{}'", query.length());
+        query = SalesforceQueryUtil.createSObjectIdQuery(query);
+      }
+      BatchInfo[] batches = runBulkQuery(bulkConnection, query, enablePKChunk);
+      LOG.debug("Number of batches received from Salesforce: '{}'", batches.length);
+      return batches;
+    } catch (AsyncApiException | IOException e) {
+      throw new RuntimeException("There was issue communicating with Salesforce", e);
+    }
+  }
+
+  /**
+   * Start batch job of reading a given guery result.
+   *
+   * @param bulkConnection bulk connection instance
+   * @param query a SOQL query
+   * @param enablePKChunk enable PK Chunk
+   * @return an array of batches
+   * @throws AsyncApiException  if there is an issue creating the job
+   * @throws IOException failed to close the query
+   */
+  public static BatchInfo[] runBulkQuery(BulkConnection bulkConnection, String query, boolean enablePKChunk)
+    throws AsyncApiException, IOException {
+
+    SObjectDescriptor sObjectDescriptor = SObjectDescriptor.fromQuery(query);
+    JobInfo job = SalesforceBulkUtil.createJob(bulkConnection, sObjectDescriptor.getName(), OperationEnum.query, null);
+    BatchInfo batchInfo;
+    try (ByteArrayInputStream bout = new ByteArrayInputStream(query.getBytes())) {
+      batchInfo = bulkConnection.createBatchFromStream(job, bout);
+    }
+
+    if (enablePKChunk) {
+      LOG.info("PKChunk is enabled, returning something now");
+      return waitForBatchChunks(bulkConnection, job.getId(), batchInfo.getId());
```

Comment:
Change "return something now" to something like 'waiting for batch chunks. Also, this and the next log msg can logged at debug log level

Reply:
Oops, I meant to remove this. Had this only for logging